### PR TITLE
Adds ability to create push items for VMIs from Errata

### PR DESCRIPTION
**Pull request by jajreidy**
Allows Errata to be a source when uploading images to clouds. @JAVGan @lslebodn @ashwini3326 PTAL

### CODE REVIEW

File: docs/sources/errata.rst [hunk 1]
```diff
@@ -9,6 +9,7 @@ Supported content types:
 * RPMs
 * Container images
 * Advisories
+* VMIs
```
Comment (JAVGan) [inline docs/sources/errata.rst:12]: Can you please change it to `Virtual machine images`?

Just to be consistent with [KojiSource](https://github.com/release-engineering/pushsource/blob/master/docs/sources/koji.rst)

```suggestion
* Virtual machine images
```

File: src/pushsource/_impl/backend/errata_source/errata_source.py [hunk 2]
```diff
@@ -247,6 +251,22 @@ def _push_items_from_container_manifests(self, erratum, docker_file_list):
...
 
+    def _push_items_from_appliance_image_list(self, appliance_image_list):
+        if not appliance_image_list:
+            return []
+
+        nvr_list = []
+        out = []
+        for image in appliance_image_list:
+            for nvr in image.keys():
+                nvr_list.append(nvr)
```
Comment (JAVGan) [inline src/pushsource/_impl/backend/errata_source/errata_source.py:262]: Just a suggestion to simplify this:

```suggestion
        for image in appliance_image_list:
            nvr_list.extend([nvr for nvr in image.keys()])
```

File: src/pushsource/_impl/backend/errata_source/errata_source.py [hunk 3]
```diff
@@ -247,6 +251,22 @@ def _push_items_from_container_manifests(self, erratum, docker_file_list):
 
         return out
 
+    def _push_items_from_appliance_image_list(self, appliance_image_list):
+        if not appliance_image_list:
+            return []
```
Comment (JAVGan) [inline src/pushsource/_impl/backend/errata_source/errata_source.py:256]: Another way around is to simply return `out` (which should be an empty list when not appliance_image_list)

```
        out = []
        
        if appliance_image_list:
            ...  # Do the for loop and fetch the push_items

        return out
```

**Review by JAVGan — COMMENTED** [review]
Just minor comments

File: src/pushsource/_impl/backend/errata_source/errata_source.py (hunk 3 above)
Comment (jajreidy) [reply src/pushsource/_impl/backend/errata_source/errata_source.py]: updated

File: src/pushsource/_impl/backend/errata_source/errata_source.py (hunk 2 above)
Comment (jajreidy) [reply src/pushsource/_impl/backend/errata_source/errata_source.py]: updated

File: docs/sources/errata.rst (hunk 1 above)
Comment (jajreidy) [reply docs/sources/errata.rst]: updated

**Review by JAVGan — APPROVED** [review]
LGTM

File: tests/errata/test_errata_vmis.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+    fake_koji.insert_archives(archives=archives, build_nvr=nvr)
+
+    with source:
+        items = list(source)
+
+    vmi_items = [i for i in items if isinstance(i, VMIPushItem)]
+
+    for push_item in vmi_items:
+        assert isinstance(push_item, AmiPushItem)
```
Comment (rohanpm) [inline tests/errata/test_errata_vmis.py:62]: It seems like at least one of these tests should be doing a bit more in-depth test of the generated push items, rather than just `isinstance`.

This test would pass even if the errata source generated AmiPushItem instances with nonsense in all the fields, wouldn't it?

File: src/pushsource/_impl/model/erratum.py [hunk 5]
```diff
@@ -354,6 +354,14 @@ class ErratumPushItem(PushItem):
     :type: list[str]
     """
 
+    appliance_image_list = attr.ib(
```
Comment (rohanpm) [inline src/pushsource/_impl/model/erratum.py:357]: I'm not convinced that this should be added as a public field on ErratumPushItem.

Do you have any plan to make use of it?  If so, can you share some details about it?  If no, please leave it out.  The info is anyway implied by the set of items you get back along with this item.

The problems I have with this field as currently defined (being a list of koji NVR):

(1) API shouldn't be unnecessarily tied to koji.

ErratumPushItem and VMIPushItem can already be generated without involving ET and koji at all (using a staging directory).  What would be put in this field in that case?  If it would be left blank - that means consumers have to be prepared to deal with a subtle difference between ErratumPushItem loaded from ET, and loaded from a staging dir, and they probably won't be aware of that.

(2) It's not consistent with or useful in the same way the other similar fields (pkglist for RPMs, and container_list for images) are useful.

Those other fields expose data which ends up going into customer portal and which makes sense to the end-user.  e.g. the names and checksums of RPMs makes sense, and the pull specs and manifest digests of container images make sense.  The NVRs of builds in koji don't make sense since that's an internal system the customers aren't aware of, don't have access to, and potentially could be replaced with something else.

TL;DR if you don't have an actual need for this, please don't add it; and if you *do* have a need for it then we need to discuss further what the format should be, because a list of koji NVRs is probably not good enough.

**Review by rohanpm — CHANGES_REQUESTED** [review]

File: src/pushsource/_impl/model/erratum.py (hunk 5 above)
Comment (jajreidy) [reply src/pushsource/_impl/model/erratum.py]: I thought it was needed but it's not so I removed it.

File: tests/errata/test_errata_vmis.py (hunk 4 above)
Comment (jajreidy) [reply tests/errata/test_errata_vmis.py]: Added more checks to it, was able to see I wasn't setting origin correctly. So it's been updated.

**Review by rohanpm — APPROVED** [review]

**Comment by rohanpm** [comment]
CI is having a problem with rpm-py-installer.  This is a cursed dependency which causes failures from time to time.  I'll have a look at it.

**Comment by rohanpm** [comment]
> CI is having a problem with rpm-py-installer. This is a cursed dependency which causes failures from time to time. I'll have a look at it.

Looks like we're going to need https://github.com/release-engineering/pushsource/pull/332 submitted first, and then this rebased.

File: tests/errata/test_errata_vmis.py [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+                arch="x86_64",
+                respin=1,
+                version="8.7",
+                base_product=None,
+                base_version=None,
+                variant=None,
+                type=None,
+            ),
+            description="",
+            type=None,
```
[the remark this answers is not in the document]
Comment (JAVGan) [reply tests/errata/test_errata_vmis.py:104]: Yes, it's being defined in StArMap for our pushes

**Comment by rohanpm** [comment]
CI is supposed to be fixed now.  Could you please rebase this change so we can get a clean CI run?

**Comment by jajreidy** [comment]
> CI is supposed to be fixed now. Could you please rebase this change so we can get a clean CI run?

Done
